graph/src/util/timed_rw_lock.rs

```diff
@@ -0,0 +1,46 @@
+use slog::{warn, Logger};
+use std::{
+    sync::RwLock,
+    time::{Duration, Instant},
+};
+
+/// Adds instrumentation for timing the performance of the lock.
+pub struct TimedRwLock<T> {
+    id: String,
+    lock: RwLock<T>,
+}
+
+impl<T> TimedRwLock<T> {
+    pub fn new(id: impl Into<String>, x: T) -> Self {
+        TimedRwLock {
+            id: id.into(),
+            lock: RwLock::new(x),
+        }
+    }
+
+    pub fn write(&self, logger: &Logger) -> std::sync::RwLockWriteGuard<T> {
+        let start = Instant::now();
+        let guard = self.lock.write().unwrap();
+        let elapsed = start.elapsed();
+        if elapsed > Duration::from_millis(100) {
```

Comment:
It would be better to make the threshold an environment variable - that way, if it turns out that this is spamming the logs, we can raise the threshold and disable logging with a simple config change.